File: templates/python/v1/composite/nodes.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Protocol


class Node(Protocol):
    def render(self) -> str:
        ...


@dataclass(slots=True)
class TextNode:
    text: str

    def render(self) -> str:
        return self.text


@dataclass(slots=True)
class ElementNode:
    tag: str
    children: List[Node] = field(default_factory=list)

    def add(self, child: Node) -> "ElementNode":
        self.children.append(child)
        return self

    def render(self) -> str:
        body = "".join(child.render() for child in self.children)
        return f"<{self.tag}>{body}</{self.tag}>"

File: templates/javascript/v1/composite/nodes.js
export class TextNode {
  constructor(text) {
    this.text = text;
  }

  render() {
    return this.text;
  }
}

export class ElementNode {
  constructor(tag) {
    this.tag = tag;
    this.children = [];
  }

  add(child) {
    this.children.push(child);
    return this;
  }

  render() {
    return `<${this.tag}>${this.children.map((c) => c.render()).join("")}</${this.tag}>`;
  }
}
